SampleViewer/DownloadsManager.cpp

```diff
@@ -0,0 +1,1016 @@
+  for (int j = 0; j < dataItemCount; ++j)
+  {
+    DataItem* dataItem = sample->dataItems()->at(j);
+    if (!dataItem->exists())
+    {
+      m_dataItems.enqueue(dataItem);
+    }
+  }
+
+  downloadNextDataItem();
+}
+
+void DownloadsManager::cancelAllDownloads()
+{
+  // Clear the queue to prevent new downloads from starting
+  m_dataItems.clear();
+
+  // Set cancel flag to prevent downloadNextDataItem from processing more
+  setCancelDownload(true);
+
+  if (m_dataItemProgress.isEmpty())
+  {
+    // This function perfroms checks for the cancel flag
+    downloadNextDataItem();
+  }
+}
+
+bool DownloadsManager::deleteAllOfflineData()
+{
+  if (!m_samples)
+  {
+    qWarning() << "DownloadsManager::deleteAllOfflineData: samples is null!";
+    return false;
+  }
+
+  const int totalSamples = m_samples->size();
+  bool success = true;
+
+  for (int i = 0; i < totalSamples; ++i)
+  {
+    Sample* sample = m_samples->at(i);
+    const int dataItemCount = sample->dataItems()->size();
+    for (int j = 0; j < dataItemCount; ++j)
+    {
+      DataItem* dataItem = sample->dataItems()->at(j);
+      deleteDataItemFile(dataItem);
+    }
+  }
+
+  updateOfflineDataProjects();
+
+  return success;
+}
+
+bool DownloadsManager::deleteProjectOfflineData(const QString& sampleName)
+{
+  if (!m_samples)
+  {
+    qWarning() << "DownloadsManager::deleteProjectOfflineData: samples is null!";
+    return false;
+  }
+
+  if (sampleName.isEmpty())
+  {
+    qWarning() << "DownloadsManager::deleteProjectOfflineData: sampleName is empty!";
+    return false;
+  }
+
+  const int totalSamples = m_samples->size();
+  bool success = false;
+
+  for (int i = 0; i < totalSamples; ++i)
+  {
+    Sample* sample = m_samples->at(i);
+    if (sample->name().toString() == sampleName)
+    {
+      const int dataItemCount = sample->dataItems()->size();
+      for (int j = 0; j < dataItemCount; ++j)
+      {
+        DataItem* dataItem = sample->dataItems()->at(j);
+        deleteDataItemFile(dataItem);
+      }
+      success = true;
+      break;
+    }
+  }
+
+  // Refresh the offline data projects list
+  updateOfflineDataProjects();
+
+  return success;
+}
+
+bool DownloadsManager::hasOfflineData(const QString& sampleName)
+{
+  if (!m_samples)
+  {
+    qWarning() << "DownloadsManager::hasOfflineData: samples is null!";
+    return false;
+  }
+
+  if (sampleName.isEmpty())
+  {
+    qWarning() << "DownloadsManager::hasOfflineData: sampleName is empty!";
+    return false;
+  }
+
+  const int totalSamples = m_samples->size();
+
+  for (int i = 0; i < totalSamples; ++i)
+  {
+    Sample* sample = m_samples->at(i);
+    if (sample->name().toString() == sampleName)
+    {
+      const int dataItemCount = sample->dataItems()->size();
+      for (int j = 0; j < dataItemCount; ++j)
+      {
+        DataItem* dataItem = sample->dataItems()->at(j);
+        if (dataItem->exists())
+        {
+          return true;
+        }
+      }
+      return false;
+    }
```

Comment:
```suggestion
    if (sample->name().toString() != sampleName)
    {
      return;
    }
    
    const int dataItemCount = sample->dataItems()->size();
    for (int j = 0; j < dataItemCount; ++j)
    {
      DataItem* dataItem = sample->dataItems()->at(j);
      if (dataItem->exists())
      {
        return true;
      }
    }
    return false;
```

I think the same applies for here. Please double check this as I just did this in the comment and didn't verify it worked.